### PR TITLE
thor: Implement a gic abstraction and add support for gic v3

### PR DESCRIPTION
**Pull request by Qwinci**
Implement a GIC abstraction and add support for GICv3 (doesn't yet support extended interrupts though).

### CODE REVIEW

File: kernel/thor/arch/arm/gic_v3.cpp [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+	uint64_t igrpen1;
+	asm volatile("mrs %0, icc_igrpen1_el1" : "=r"(igrpen1));
+	igrpen1 |= 1;
+	asm volatile("msr icc_igrpen1_el1, %0" : : "r"(igrpen1));
+
+	for(int i = 0; i < 32; ++i) {
+		auto pin = static_cast<GicPinV3 *>(gicV3->getPin(i));
+		pin->mask();
+		pin->setPriority_(defaultPrio);
+		// todo? irq 28
```
Comment (qookei) [inline kernel/thor/arch/arm/gic_v3.cpp:357]: What's this about?

File: kernel/thor/arch/arm/thor-internal/arch/gic.hpp [hunk 2]
```diff
@@ -7,92 +7,42 @@
...
-	uint8_t ifaceNo_;
+	virtual Pin *setupIrq(uint32_t irq, TriggerMode trigger) = 0;
+	virtual Pin *getPin(uint32_t irq) = 0;
 };
 
 initgraph::Stage *getIrqControllerReadyStage();
 
 void initGicOnThisCpu();
 
+extern Gic* gic;
```
Comment (qookei) [inline kernel/thor/arch/arm/thor-internal/arch/gic.hpp:46]: Pointer on the wrong side :^)

File: kernel/thor/arch/arm/gic_v3.cpp [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+		arch::scalar_store_relaxed(dist->space_, dist_reg::irqConfigBase + offset, v);
+
+		auto group = arch::scalar_load_relaxed<uint32_t>(dist->space_, dist_reg::irqGroupBase + groupOffset);
+		group |= 1U << groupBitOffset;
+		arch::scalar_store_relaxed(dist->space_, dist_reg::irqGroupBase + groupOffset, group);
+
+		auto groupMod = arch::scalar_load_relaxed<uint32_t>(dist->space_, dist_reg::irqGroupModBase + groupOffset);
+		groupMod &= ~(1U << groupBitOffset);
+		arch::scalar_store_relaxed(dist->space_, dist_reg::irqGroupModBase + groupOffset, groupMod);
+	}
```
[the remark this answers is not in the document]
Comment (Qwinci) [reply kernel/thor/arch/arm/gic_v3.cpp]: Yeah, they only have different base offset + space. I changed it so they use the same code but with different space (subspaced in the redist case).

File: kernel/thor/arch/arm/gic_v3.cpp (hunk 1 above)
Comment (Qwinci) [reply kernel/thor/arch/arm/gic_v3.cpp]: The todo was accidentally left in, I removed it.

**Review by qookei — APPROVED** [review]

**Pull request merged**